### PR TITLE
Additional functional issues

### LINKED ISSUE #249: [IDO pallet] Several extrinsics do not update the value of storage items for tracking funds, causing inconsistency in the runtime state
Link: https://github.com/Polkadex-Substrate/audit-sr-labs/issues/9 

Summary
In the IDO pallet, the withdraw_token, withdraw_raise and participate_in_round do not update the storage items tracking the amounts of funds. This creates inconsistencies in the runtime state that could lead to financial loss for users.

Issue details
The withdraw_token does not update the value of actual_raise, enabling a creator to call this extrinsic multiple times. No matter how many times the creator calls this extrinsic, the actual_raise value for the funding round would remain unchanged. This enables the creator to transfer remaining_token amount of tokens to a beneficiary arbitrary amount of times.

Similarly, the withdraw_raise extrinsic does not check whether the raise of the funding round was already withdrawn or not. A creator can call this extrinsic over and over and transfer total_raise amount of tokens to a beneficiary arbitrary amount of times.

The extrinsic participate_in_round also suffers from a similar issue. If an investor calls it for the second time, it will deposit the funds from his account but it will replace his previous share info (saved in InvestorShareInfo). If an investor calls this extrinsic the second time, they will lose min(amount_1, amount_2) amount of tokens considering the two extrinsic calls.

Risk
These inconsistencies may lead to financial loss in tokens for users. Such bugs could also undermine the reputation and trust in the system.

Mitigation
We suggest to keep the value of these storage items updated in the respective extrinsics and and check the value before performing any balance transfer. Consider limiting the number of times (to for example, only once) these extrinsics can be called.

### LINKED ISSUE #253: [IDO Pallet] Additional functional issues
https://github.com/Polkadex-Substrate/audit-sr-labs/issues/13

Summary
During the review of the IDO pallet (Polkadex/pallets/Polkadex_Ido/src/lib.rs) we also uncovered several functional issues. Since these issues are not security issues by nature, we did not assign a severity and associated risk with them. Please find the details as follows.

Details
Missing checks on the IDO funding round status. The extrinsics withdraw_token and withdraw_raise do not check the current block number, in the context of whether the IDO round has already finished. This could lead to an issue where IDO creators withdraw the raised amount before the IDO round finishes.

Missing implementation logic for investor token withdrawal. In the current implementation, investors are not paid out after calling claim_tokens; the amount of tokens they should get is simply stored in InfoClaimAmount<T>.

Incorrect sanity check in participate_in_round. The following sanity check is in place to ensure that an investor is not investing more than their max_amount in an IDO round. However, in the current implementation, the relation check is performed the other way around:

ensure!(amount >= max_amount, Error::<T>::NotAValidAmount);
instead of

ensure!(amount <= max_amount, Error::<T>::NotAValidAmount);

**Fredrik Edit:**
Create a unit test  to catch this error

### LINKED ISSUE #290: Funding and selling token can be same in polkadexIDO.
**we can create a round where the token bought and sold are the same**
eg : it is possible to raise funds in PDEX while the token sold  by that round is also PDEX
(token_a and token_b same).

### PR DESCRIPTION
**Pull request by mambisi**
Fixes #253 
Fixes #290 
Fixes #249

### CODE REVIEW

File: pallets/Polkadex_Ido/src/test.rs [hunk 1]
```diff
@@ -548,6 +593,16 @@ fn test_withdraw_token() {
...
+
+        //Test creator withdraw tokens during when fundraising is not close should return error : WithdrawalBlocked
+        system::Pallet::<Test>::set_block_number(closing_block_number - 1);
+        assert_noop!(
+            PolkadexIdo::withdraw_token(Origin::signed(ALICE), round_id, investor_address),
+            Error::<Test>::WithdrawalBlocked
+        );
+
+        //Test creator withdraw tokens during when fundraising is closed should be successful
+        system::Pallet::<Test>::set_block_number(closing_block_number + 1);
```
Comment (simonsso) [inline pallets/Polkadex_Ido/src/test.rs:605]: Set this to closing_block_number to test corner case.

**Review by simonsso — CHANGES_REQUESTED** [review]

File: pallets/Polkadex_Ido/src/lib.rs [hunk 2]
```diff
@@ -411,17 +415,32 @@ decl_module! {
             ensure!(<InfoFundingRound<T>>::contains_key(&round_id.clone()), Error::<T>::FundingRoundDoesNotExist);
             let current_block_no = <frame_system::Pallet<T>>::block_number();
             let funding_round = <InfoFundingRound<T>>::get(round_id);
-            if current_block_no >= funding_round.start_block {
-                let investor_share = <InvestorShareInfo<T>>::get(round_id, investor_address.clone());
-                let total_released_block: T::BlockNumber = current_block_no - funding_round.start_block;
-                let tokens_released_for_given_investor: T::Balance = Self::block_to_balance(total_released_block)
+            ensure!(current_block_no >= funding_round.close_round_block, Error::<T>::WithdrawalBlocked);
```
[the remark this answers is not in the document]
Comment (mambisi) [reply pallets/Polkadex_Ido/src/lib.rs:418]: From the SR lab audit the suggestion was to claim token after the round ended

File: pallets/Polkadex_Ido/src/lib.rs [hunk 3]
```diff
@@ -411,17 +415,32 @@ decl_module! {
...
-                <LastClaimBlockInfo<T>>::insert(round_id, investor_address.clone(), current_block_no);
-            }
+                //Check if investor previously claimed the tokens
+            let claimed_tokens = if <InfoClaimAmount<T>>::contains_key(&round_id, &investor_address) {
+                <InfoClaimAmount<T>>::get(&round_id,&investor_address)
+                }else {
+                    Zero::zero()
+                };
+                // claimable_tokens : is the total amount of token the investor can withdraw(claim)  in their account
+            let claimable_tokens = total_tokens_released_for_given_investor - claimed_tokens;
```
[the remark this answers is not in the document]
Comment (mambisi) [reply pallets/Polkadex_Ido/src/lib.rs:436]: Ok

File: pallets/Polkadex_Ido/src/lib.rs [hunk 4]
```diff
@@ -411,17 +415,32 @@ decl_module! {
...
+            T::Currency::transfer(funding_round.token_a, &round_account_id, &investor_address, claimable_tokens);
+
+            <InfoClaimAmount<T>>::insert(round_id, investor_address.clone(), total_tokens_released_for_given_investor);
+                // TODO : remove
```
[the remark this answers is not in the document]
Comment (mambisi) [reply pallets/Polkadex_Ido/src/lib.rs:440]: The line below it serve no purpose, i dont see being used anywhere else in the code

File: pallets/Polkadex_Ido/src/lib.rs [hunk 5]
```diff
@@ -520,16 +542,22 @@ decl_module! {
...
             ensure!(<InfoFundingRound<T>>::contains_key(&round_id), Error::<T>::FundingRoundDoesNotExist);
             ensure!(<InfoProjectTeam<T>>::contains_key(&creator), <Error<T>>::CreaterDoesNotExist);
             let info_round_id = <InfoProjectTeam<T>>::get(&creator);
             ensure!(info_round_id.eq(&round_id), <Error<T>>::NotACreater);
             let funding_round = <InfoFundingRound<T>>::get(round_id);
-            let total_raise = funding_round.amount.saturating_mul(funding_round.token_a_priceper_token_b);
-            let remaining_token = total_raise.saturating_sub(funding_round.actual_raise);
-            <T as Config>::Currency::transfer(AssetId::POLKADEX, &creator, &beneficiary, remaining_token)?;
+            ensure!(current_block_no >= funding_round.close_round_block, Error::<T>::WithdrawalBlocked);
+            let total_tokens_bought_by_investors = funding_round.actual_raise.saturating_mul(1_u32.saturated_into::<T::Balance>()/ funding_round.token_a_priceper_token_b);
```
[the remark this answers is not in the document]
Comment (mambisi) [reply pallets/Polkadex_Ido/src/lib.rs:554]: You have to manually format code that in a marco

File: pallets/Polkadex_Ido/src/test.rs (hunk 1 above)
Comment (mambisi) [reply pallets/Polkadex_Ido/src/test.rs:605]: OK

File: pallets/Polkadex_Ido/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (simonsso) [reply pallets/Polkadex_Ido/src/lib.rs:440]: The data we read in 

                let claimed_tokens = if <InfoClaimAmount<T>>::contains_key(&round_id, &investor_address) {
                <InfoClaimAmount<T>>::get(&round_id,&investor_address)

should be updated here

File: pallets/Polkadex_Ido/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (simonsso) [reply pallets/Polkadex_Ido/src/lib.rs:418]: It is correct, I just put the comment to remind future us that we did check it.

**Review by simonsso — APPROVED** [review]
